### ISSUE TITLE
Require dependencies for custom.js

### ISSUE THREAD

**Issue by dsblank** [ipython/ipython#8265]
In IPython2 one could put this in one's custom.js and it would work:

``` javascript
IPython.load_extensions('calico-spell-check'); 
```

In IPython 3, one needs:

``` javascript
require([ 
     'base/js/namespace', 
     'base/js/events' 
], function(IPython, events) { 
     events.on('app_initialized.NotebookApp', function(){ 
         IPython.load_extensions('calico-spell-check'); 
     }); 
 });
```

Can't these dependencies by required before attempting to run custom.js?

See also conversation at https://github.com/ipython/ipython/issues/3216#issuecomment-89668489

**Comment by minrk** [comment]
They should be, and they used to be. I don't know if something changed in how require resolves dependencies or what.

**Comment by Carreau** [comment]
No, and this is how javascript works, we put custom as late as possible on dependencies list, but there always was a small race condition. There is no difference between **reading** the file and **executing** the content of the file. This is what libraries like require try to go around with module definition like : 

```
define(...)
```

To actually make the module variable evaluation deferred until the dependencies of the module are loaded. This is actually pretty complex to handle especially when things have circular dependencies.


Also this not that much different from Python, where module usually don't have side effect on import, and you write a `if __name__== '__main__'` in the end which is a clear gross hack. What is above is the same.

There is also no (easy) way in require to express a depencies in a file which is not yet loaded. 
Require expect to be able to read a file and figure out dependencies before the file actually does anything.


It's like looking at the schrodinger cat in the box. Once you look it's dead, unless you hooked him up to a machine to keep it alive.

**Comment by minrk** [comment]
@Carreau and yet it used to work during 3.x dev, and we broke it.

**Comment by takluyver** [comment]
Couldn't we simply load it with a second require call in the callback that fires when all of the regular dependencies are loaded?

For loading extensions, you can also specify them declaratively in frontend config; this is applied after app_initialized fires. See e.g. [the cite2c install script](https://github.com/takluyver/cite2c/blob/master/install.py) for how I do this.

**Comment by minrk** [comment]
@takluyver yes, we could.

**Comment by dsblank** [comment]
It would be great to target a fix for ipython-3.2. How easy is this to fix?
